File: zig/isbn-verifier/isbn_verifier.zig
const std = @import("std");
var gpa = std.heap.GeneralPurposeAllocator(.{}){};

/// Returns a bool dependant on if `s` is a valid ISBN
pub fn isValidIsbn10(s: []const u8) bool {
    const length = s.len;
    if (length < 10 or length > 13) return false;

    var multiplier: u8 = 10;
    var sum: u32 = 0;
    
    for (s) |c| {
        var digit: u8 = undefined;
        switch (c) {
            '-' => continue,
            '0'...'9' => digit = c - '0',
            'X' => digit = 10,
            else => return false,
        }

        if (multiplier < 1 or digit == 10 and multiplier > 1) return false;

        sum += (digit * multiplier);
        multiplier -= 1;
    }

    return multiplier == 0 and sum % 11 == 0;
}

File: go/isbn-verifier/isbn_verifier.go
package isbn

func IsValidISBN(isbn string) bool {
	if len(isbn) < 10 || len(isbn) > 13 {
		return false
	}

	multiplier := 10
	checksum := 0
	digit := 0
	for _, val := range isbn {
		switch val {
		case '-':
			continue
		case '0', '1', '2', '3', '4', '5', '6', '7', '8', '9':
			digit = int(val - '0')
		case 'X':
			digit = 10
		default:
			return false
		}

		if multiplier < 1 || digit == 10 && multiplier > 1 {
			return false
		}
		checksum += digit * multiplier
		multiplier -= 1
	}
	return multiplier == 0 && checksum%11 == 0
}
